### PR TITLE
retryGo refactor

### PR DESCRIPTION
**Pull request by vponline**
This PR does the following:
~~1. Incorporates Airnode's array `GoResult` type~~
~~2. Includes Airnode's `retryOperation` and `promiseTimeout` retry logic into promise-utils `go` function~~
1. Refactors `go` and adds retry functionality
2. Adds Airkeeper's `retryGo` function
3. Adds `timeoutGo` and `retryTimeoutGo`

### CODE REVIEW

**Comment by amarthadan** [comment]
Oh, I didn't expect that you'll start implementing the tuple version of the `go` function :sweat_smile: We've decided that we'll switch to returning [an object](https://github.com/api3dao/promise-utils/pull/2#discussion_r812124852). Sorry, I could have mentioned that :confused:

File: src/index.ts [hunk 1]
```diff
@@ -27,15 +30,27 @@ const createGoError = <E extends Error>(err: unknown): GoResultError<E> => {
   return fail(new Error('' + err));
 };
 
-export const go = async <T, E extends Error>(fn: Promise<T> | (() => Promise<T>)): Promise<GoResult<T, E>> => {
-  // We need try/catch because `fn` might throw sync errors as well
+export const go = async <T, E extends Error>(
+  fn: () => Promise<T>,
```
Comment (vponline) [inline src/index.ts:34]: > Oh, I didn't expect that you'll start implementing the tuple version of the `go` function sweat_smile We've decided that we'll switch to returning [an object](https://github.com/api3dao/promise-utils/pull/2#discussion_r812124852). Sorry, I could have mentioned that confused Can you remove that part? I'll review the rest (the retry functionality) once that's gone.

No problem, I had a feeling this might be the case :sweat_smile:

Fixed now and I think the only change major change would be to `fn` here.

File: src/index.ts [hunk 2]
```diff
@@ -56,3 +71,89 @@ export function assertGoError<E extends Error>(result: GoResult<any, E>): assert
     throw new Error('Assertion failed. Expected error, but no error was thrown');
   }
 }
+
+// Adapted from:
+// https://github.com/then/is-promise
+export function isPromise(obj: any) {
+  return !!obj && (typeof obj === 'object' || typeof obj === 'function') && typeof obj.then === 'function';
+}
```
Comment (amarthadan) [inline src/index.ts:79]: Do we need this for anything?

File: src/index.ts (hunk 1 above)
Comment (amarthadan) [inline src/index.ts:34]: Why did you get rid of the `Promise<T>` type? You can always wrap it later in the function but it's quite handy if you can pass the promise directly without wrapping it when using the library.

File: src/index.ts [hunk 3]
```diff
@@ -27,15 +30,27 @@ const createGoError = <E extends Error>(err: unknown): GoResultError<E> => {
...
+    if (options?.timeoutMs) {
+      return promiseTimeout(options.timeoutMs, fn())
         .then(success)
         .catch((err) => createGoError(err));
     }
-    return fn.then(success).catch((err) => createGoError(err));
+
+    return fn()
+      .then(success)
+      .catch((err) => createGoError(err));
```
Comment (amarthadan) [inline src/index.ts:53]: Why treat this as 3 different cases? As far as I can tell you can always pass it through the `retry` function. With zero retries and no timeout if needed. I understand that the library has a bit of strange behavior, stopping retrying once the timeout is reached. But I think that can be handled differently. If we have both multiple retries and timeout and we run out of time, we can just run the whole thing again with one less retry. I haven't actually tried to write the code but I think it might work.

File: src/index.ts [hunk 4]
```diff
@@ -56,3 +71,89 @@ export function assertGoError<E extends Error>(result: GoResult<any, E>): assert
...
+    }
+
+    return run();
+  });
+
+  return promiseTimeout(maxTimeoutMs, promise);
+}
+
+export const retryGo = <T>(fn: () => Promise<T>, options?: PromiseOptions) =>
+  go(() => retryOnTimeout(DEFAULT_RETRY_TIMEOUT_MS, fn), options);
```
Comment (amarthadan) [inline src/index.ts:159]: I think we should provide 4 functions:
* `go` - 0 retries, no timeout
* `retryGo` - same as `go` with a default number of retries, let's say 3
* `timeoutGo` - same as `go` with a default timeout, let's say 10s
* `retryTimeoutGo` - same as `go` with a default number of retries (3) and a default timeout (10s)

And you can pass the options object to all of them to override the defaults.

File: src/index.ts [hunk 5]
```diff
@@ -1,7 +1,10 @@
+import { AttemptOptions, retry, sleep } from '@lifeomic/attempt';
+const DEFAULT_RETRY_DELAY_MS = 50;
```
Comment (amarthadan) [inline src/index.ts:2]: I don't think we need a default delay. Let's keep it as an option if somebody wants to set it but I'd use the library's default otherwise.

**Review by amarthadan — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 2 above)
Comment (vponline) [reply src/index.ts:103]: It doesn't seem to be used in Airnode or Airkeeper, remove it?

File: src/index.ts [hunk 6]
```diff
@@ -56,3 +95,96 @@ export function assertGoError<E extends Error>(result: GoResult<any, E>): assert
...
+          // If the error is NOT a timeout error, then we reject immediately
+          return reject(reason);
+        });
+    }
+
+    return run();
+  });
+
+  return promiseTimeout(maxTimeoutMs, promise);
+}
```
Comment (vponline) [inline src/index.ts:180]: Do we need these functions at all?

File: src/index.ts (hunk 2 above)
Comment (amarthadan) [reply src/index.ts:103]: Yes, please, do so :slightly_smiling_face:

File: src/index.ts [hunk 7]
```diff
@@ -27,15 +28,53 @@ const createGoError = <E extends Error>(err: unknown): GoResultError<E> => {
...
+    delay: options?.retryDelayMs || 200,
+    maxAttempts: options?.retries || 1,
+    initialDelay: 0,
+    minDelay: 0,
+    maxDelay: 0,
+    factor: 0,
+    timeout: options?.timeoutMs || 0,
+    jitter: false,
+    handleError: null,
+    handleTimeout: options?.timeoutMs
```
Comment (vponline) [inline src/index.ts:45]: We could handle the retry + timeout options case through the library's `handleTimeout` function like this.

The difference will be that if options.timeoutMs is defined and it ends up retrying the maximum times and still fails, it will throw an `Operation timed out after final retry` instead of re-throwing the original error (if any).

Is this more like what you had in mind?

File: src/index.test.ts [hunk 8]
```diff
@@ -63,6 +84,72 @@ describe('basic go usage', () => {
...
+
+  it('retries and resolves unsuccessful asynchronous functions', async () => {
+    jest
+      .spyOn(operations, 'errorFn')
+      .mockRejectedValueOnce(new Error('Operation timed out, retries left: 1'))
+      .mockRejectedValueOnce(new Error('Operation timed out after final retry'));
+
+    const res = await retryTimeoutGo(operations.errorFn, { timeoutMs: 100, retries: 3 });
+    expect(operations.errorFn).toHaveBeenCalledTimes(3);
+    expect(res).toEqual(fail(new Error('Operation timed out after final retry')));
```
Comment (vponline) [inline src/index.test.ts:149]: Note the different error here compared to using `retryGo` [here](https://github.com/api3dao/promise-utils/pull/5/files#diff-258035e5968f6bf645400d417f310218d7d9a9a10606a3c34e7f55db193f58f3R113)

File: src/index.ts (hunk 7 above)
Comment (amarthadan) [reply src/index.ts:54]: Yes and no :smile: 

I definitely don't want to use [exceptions for a flow control](https://softwareengineering.stackexchange.com/a/189225). Would it be problematic to call the whole function again from within the `handleTimeout` handler? I mean the `go` function. I think the flow ends after the `handleTimeout` anyway, so it shouldn't be a problem. From the documentation:
```
The handleTimeout function should return a Promise that will be the return value of the retry() function.
```
So, the result will be eventually returned. I think that can work.

File: src/index.ts [hunk 9]
```diff
@@ -56,3 +104,90 @@ export function assertGoError<E extends Error>(result: GoResult<any, E>): assert
...
+}
+
+export const retryGo = <T>(fn: Promise<T> | (() => Promise<T>), options?: PromiseOptions) =>
+  go(fn, { retries: 3, ...options });
+
+export const timeoutGo = <T>(fn: Promise<T> | (() => Promise<T>), options?: PromiseOptions) =>
+  go(fn, { timeoutMs: 10_000, ...options });
+
+export const retryTimeoutGo = <T>(fn: Promise<T> | (() => Promise<T>), options?: PromiseOptions) => {
+  return go(fn, { retries: 3, timeoutMs: 10_000, ...options });
```
Comment (amarthadan) [inline src/index.ts:192]: Please, use constants for those defaults.

File: src/index.ts (hunk 7 above)
Comment (vponline) [reply src/index.ts:45]: Yes it is possible to call `go` inside `handleTimeout`, but it seems to work a bit differently (see [here](https://github.com/api3dao/promise-utils/pull/5/files#diff-258035e5968f6bf645400d417f310218d7d9a9a10606a3c34e7f55db193f58f3R115)).

Or do you see another way to use it? I think we can also remove of `retryFn` entirely in this case as well.

File: src/index.test.ts [hunk 10]
```diff
@@ -63,6 +63,83 @@ describe('basic go usage', () => {
...
+      ),
+    errorFn: () => new Promise((_res, rej) => setTimeout(() => rej(new Error('Computer says no')), 20)),
+  };
+
+  it('retries and resolves timed out functions', async () => {
+    const res = await retryTimeoutGo(operations.successFn, { timeoutMs: 50, retries: 3 });
+    expect(res).toEqual(success(2));
+  });
+
+  it('retries and resolves successful asynchronous functions', async () => {
```
Comment (vponline) [inline src/index.test.ts:120]: Added one more test for the case where `go` is called from inside the timeout. Is this what is needed?

**Review by amarthadan — CHANGES_REQUESTED** [review]
You're actually missing a few test cases. You need to test that:
* retries (alone) works
  * with no retries it fails after the first try
  * with retries set
    * emulate failures before success
    * let it fail every time and make sure that it tried multiple times
* timeout (alone) works
  * with no timeout it will either succeed or fail on its own
  * with timeout set
    * it won't affect the operation if it succeeds/fails within the timeout
    * it will timeout if the operation takes longer than the timeout
* retries and timeout together works
  * combining these won't affect use-cases above
  * operation that takes longer than the timeout will retry set number of times until it timeouts for good

You've already covered some of them, I just wanted to list all the cases I see that need to be tested.

File: src/index.ts (hunk 7 above)
Comment (vponline) [reply src/index.ts:69]: I can't commit a failing test, but I had this kind of a scenario:
```js
it('retries and resolves timed out functions', async () => {
    const res = await retryTimeoutGo(
      () =>
        new Promise((res) =>
          setTimeout(() => {
            res(2);
          }, 50)
        ),
      { timeoutMs: 50, retries: 3 }
    );
    expect(res).toEqual(fail(new Error('Operation timed out after final retry')));
  });
```

In the end this fails with the following result which doesn't seem right:
```js
{
  data: {
    data: {
      error: [Error: Operation timed out after final retry],
      success: false,
    },
    success: true,
  },
  success: true,
}
```

However, this test passed with the `retryFn` (because if was handled through the exception).

File: src/index.ts [hunk 11]
```diff
@@ -22,37 +52,60 @@ export const goSync = <T, E extends Error>(fn: () => T): GoResult<T, E> => {
...
-  if (err instanceof Error) return fail(err);
-  return fail(new Error('' + err));
-};
+export const go = async <T, E extends Error>(
+  fn: Promise<T> | (() => Promise<T>),
+  options?: PromiseOptions
+): Promise<GoResult<T, E>> => {
+  const attemptOptions: AttemptOptions<any> = {
+    delay: options?.retryDelayMs || 200,
+    maxAttempts: options?.retries || 1,
```
Comment (vponline) [inline src/index.ts:61]: One more point question about `maxAttempts`:
>The maximum number of attempts or 0 if there is no limit on number of attempts.

In Airnode this was defined as: `maxAttempts: options.retries + 1`

The default for the library is 3, but do we also want to support `0`/unlimited or the way it is in Airnode currently?

Comment (amarthadan) [reply src/index.ts:61]: > In Airnode this was defined as: maxAttempts: options.retries + 1

Right. We're using retries so the number of attempts must be 1 (the original attempt) + the number of retries. I think that's fine.

> The default for the library is 3, but do we also want to support 0/unlimited or the way it is in Airnode currently?

What do you mean by "the way it is in Airnode currently"? Do we have unlimited retries somewhere?
But I think we should have as a default for the main `go` function 0 retries, which means no retries, just the one attempt and a default of 3 retries (1 attempt + 3 retries) for the `retryGo` and `retryTimeoutGo` functions.

Comment (vponline) [reply src/index.ts:61]: >Right. We're using retries so the number of attempts must be 1 (the original attempt) + the number of retries. I think that's fine.

Makes perfect sense, thanks!

> Do we have unlimited retries somewhere?
But I think we should have as a default for the main go function 0 retries, which means no retries, just the one attempt and a default of 3 retries (1 attempt + 3 retries) for the retryGo and retryTimeoutGo functions.

No unlimited retries are used currently, just wanted to confirm they're not needed :+1:

File: src/index.ts (hunk 7 above)
Comment (vponline) [reply src/index.ts:69]: This makes perfect sense and works for that case as well, Will add

File: src/index.test.ts [hunk 12]
```diff
@@ -63,6 +63,151 @@ describe('basic go usage', () => {
...
+  it('resolves unsuccessful asynchronous functions', async () => {
+    const res = await retryTimeoutGo(operations.errorFn, { timeoutMs: 50, retries: 3 });
+    expect(res).toEqual(fail(new Error('Computer says no')));
+  });
+
+  it('retries and resolves successful asynchronous functions', async () => {
+    jest
+      .spyOn(operations, 'successFn')
+      .mockRejectedValueOnce(new Error('Operation timed out, retries left: 1'))
+      .mockRejectedValueOnce(new Error('Operation timed out after final retry'));
```
Comment (amarthadan) [inline src/index.test.ts:167]: Same here, this is not timeouting per se

File: src/index.test.ts [hunk 13]
```diff
@@ -63,6 +63,151 @@ describe('basic go usage', () => {
...
+        new Promise((res) =>
+          setTimeout(() => {
+            res(2);
+          }, 200)
+        ),
+      { timeoutMs: 50, retries: 3 }
+    );
+    expect(res).toEqual(fail(new Error('Operation timed out after final retry')));
+  });
+});
```
Comment (amarthadan) [inline src/index.test.ts:209]: This is where you should check whether the function was called the correct number of times and then it timeouted.

**Review by amarthadan — CHANGES_REQUESTED** [review]
Last round of changes, I promise :sweat_smile:

**Comment by vponline** [comment]
> Last round of changes, I promise sweat_smile

I've updated the misleading (which I also confused myself with :sweat_smile: ) errors and test names.

File: src/index.test.ts [hunk 14]
```diff
@@ -63,6 +63,147 @@ describe('basic go usage', () => {
...
+    jest
+      .spyOn(operations, 'successFn')
+      .mockRejectedValueOnce(new Error('Error 1'))
+      .mockRejectedValueOnce(new Error('Error 2'))
+      .mockRejectedValueOnce(new Error('Computer says no'));
+
+    const res = await retryTimeoutGo(operations.successFn, { timeoutMs: 100, retries: 2 });
+    expect(operations.successFn).toHaveBeenCalledTimes(3);
+    expect(res).toEqual(fail(new Error('Computer says no')));
+  });
```
Comment (amarthadan) [inline src/index.test.ts:195]: According to the name, this should end up being successful.

**Review by amarthadan — APPROVED** [review]
One test fix and it's good to go :+1: Sorry for so many back and forth :sweat_smile:

Can you please also update `tsconfig.build-es6.json` and change `moduleResolution` to `node`? It started to fail (when you run `yarn build`) when you added a dependency.

File: src/index.test.ts (hunk 14 above)
Comment (vponline) [reply src/index.test.ts]: :smile: Thanks for bearing with me. I've removed that test as the one [here](https://github.com/api3dao/promise-utils/pull/5/files#diff-258035e5968f6bf645400d417f310218d7d9a9a10606a3c34e7f55db193f58f3R163-R172) already tests that case.

**Comment by amarthadan** [comment]
Feel free to merge :slightly_smiling_face:

File: tsconfig.build-es6.json [hunk 15]
```diff
@@ -3,7 +3,7 @@
   "compilerOptions": {
     "target": "es6",
     "module": "es6",
-    "moduleResolution": "classic",
+    "moduleResolution": "node",
```
Comment (Siegrift) [inline tsconfig.build-es6.json:6]: Why was this changed?

File: src/index.ts [hunk 16]
```diff
@@ -22,37 +52,73 @@ export const goSync = <T, E extends Error>(fn: () => T): GoResult<T, E> => {
...
-// See: https://github.com/microsoft/TypeScript/issues/34523#issuecomment-542978853
-export function assertGoError<E extends Error>(result: GoResult<any, E>): asserts result is GoResultError<E> {
-  if (result.success) {
-    throw new Error('Assertion failed. Expected error, but no error was thrown');
-  }
-}
+export const timeoutGo = <T>(fn: Promise<T> | (() => Promise<T>), options?: PromiseOptions) =>
+  go(fn, { timeoutMs: DEFAULT_RETRY_TIMEOUT_MS, ...options });
+
+export const retryTimeoutGo = <T>(fn: Promise<T> | (() => Promise<T>), options?: PromiseOptions) => {
```
Comment (Siegrift) [inline src/index.ts:122]: Why there is this function with hardcoded retries?

Comment (Siegrift) [reply src/index.ts:122]: Actually do we need all of these functions? Isn't just optional argument in the go and goSync enough? cc: @amarthadan ?

File: src/index.ts [hunk 17]
```diff
@@ -22,37 +52,73 @@ export const goSync = <T, E extends Error>(fn: () => T): GoResult<T, E> => {
...
+          if (context.attemptsRemaining > 0) {
+            const res = await retryFnWrapper(fn, { ...options, maxAttempts: context.attemptsRemaining });
+
+            if (res.success) {
+              return res.data;
+            } else {
+              throw res.error;
+            }
+          }
+          throw new Error(`Operation timed out after final retry`);
```
Comment (Siegrift) [inline src/index.ts:101]: This error message will also be used when no retries are used so it's a bit connfusing. Let's just change that to `Operation timed out`.

File: src/index.ts [hunk 18]
```diff
@@ -22,37 +52,73 @@ export const goSync = <T, E extends Error>(fn: () => T): GoResult<T, E> => {
...
+  return retryFn(() => fn, attemptOptions);
+};
+
+const retryFn = async <T, E extends Error>(
+  fn: () => Promise<T>,
+  attemptOptions: AttemptOptions<T>
+): Promise<GoResult<T, E>> => {
+  return retry(fn, attemptOptions)
+    .then(success)
+    .catch((err) => {
```
Comment (Siegrift) [inline src/index.ts:71]:
```
.catch(createGoError)
```

**Review by Siegrift — COMMENTED** [review]
I like the extra parameter in the go functions, but I am not sure about the other functions with predefined retries. I'd remove those.

Also, the documentation in the README needs to be updated.

File: tsconfig.build-es6.json (hunk 15 above)
Comment (amarthadan) [reply tsconfig.build-es6.json:6]: Because the build wasn't working without it.

Comment (Siegrift) [reply tsconfig.build-es6.json:6]: But this will not work for browsers.

Comment (amarthadan) [reply tsconfig.build-es6.json:6]: I thought that the `target` and `module` parts are the important ones. Feel free to fix it.

Comment (Siegrift) [reply tsconfig.build-es6.json:6]: Actually, you are right. There is this written in the docs:

```
If not specified, the default is [Node](https://www.typescriptlang.org/docs/handbook/module-resolution.html#node) for --module commonjs, and [Classic](https://www.typescriptlang.org/docs/handbook/module-resolution.html#classic) otherwise (including when [module](https://www.typescriptlang.org/tsconfig#module) is set to amd, system, umd, es2015, esnext, etc.).
```

so I assumed that it should be `classic`, but yeah - it shoul be `node`. I guess this turned out to be an issue after we've added a dependency which couldn't be imported.

**Comment by vponline** [comment]
@Siegrift Should I make these changes?

**Comment by Siegrift** [comment]
Yeah, please do - I have too much to catch up since last two weeks :/

File: src/index.ts (hunk 18 above)
Comment (vponline) [reply src/index.ts:71]: hmm looks like we can do:
```js
.catch<GoResultError<E>>(createGoError);
```
or
```js
.catch((err) => createGoError(err));
```

Any preference? :smile:

Comment (Siegrift) [reply src/index.ts:71]: Oh, I haven't realized that TS will complain about that. I find the latter nicer, but this is not that important :)
